src/journal/journald-context.c

```diff
@@ -268,6 +273,46 @@ static int client_context_read_label(
         return 0;
 }
 
+static int client_context_read_log_filter_patterns(ClientContext *c, const char *cgroup) {
+        const char *pattern;
+        _cleanup_free_ char *xattr = NULL;
+        _cleanup_set_free_ Set *allow_list = NULL, *deny_list = NULL;
+        Set **current_list = &allow_list;
+        int r;
+
+        r = cg_get_xattr_malloc(SYSTEMD_CGROUP_CONTROLLER, cgroup, "user.journald_log_filter_patterns", &xattr);
+        if (r < 0 && r != -ENODATA)
+                return log_debug_errno(r, "Failed to get user.journald_log_filter_patterns xattr for %s: %m", cgroup);
+
+        NULSTR_FOREACH(pattern, xattr) {
+                pcre2_code *compiled_pattern = NULL;
+
+                if (streq(pattern, "\xff")) {
+                        current_list = &deny_list;
+                        continue;
+                }
+
+                r = pattern_compile_and_log(pattern, 0, &compiled_pattern);
+                if (r < 0)
+                        return log_debug_errno(r, "Failed to compile log filtering pattern '%s' for %s: %m",
+                                               pattern, cgroup);
+
+                r = set_ensure_put(current_list, &pcre2_code_hash_ops_free, compiled_pattern);
+                if (r < 0) {
+                        compiled_pattern = pattern_free(compiled_pattern);
```

Comment:
We usually use `_cleanup_` attribute and `TAKE_PTR()` on success.

Reply:
Do you suggest:

```
 _cleanup_(pattern_freep) pcre2_code *compiled_pattern = NULL;
[...]

r = set_ensure_put(current_list, &pcre2_code_hash_ops_free, compiled_pattern);
if (r < 0)
    return r;

TAKE_PTR(compiled_pattern);
```

Also,  `r = set_ensure_put(current_list, &pcre2_code_hash_ops_free, TAKE_PTR(compiled_pattern));` wouldn't work as expected if the call fails.

Reply:
Right.

Reply:
Changing to use `TAKE_PTR` then. 

Side note: by using `_cleanup_(pattern_freep)`, if we forgot to call `TAKE_PTR` on success (which could happen with complex code path and refactoring) then could have a use after free. On the other hand, if we don't use `_cleanup_(pattern_freep)` and forgot to free the pattern on error, the worst case would be to leak memory.

Reply:
Actually, just chatted with @DaanDeMeyer on this subject, he told me about `set_ensure_consume()` which is the elegant way of solving this :D

Reply:
IIRC (but may not be correct, now I am from cellphone), set_ensure_consume() calls free() on failure, not free_func provided by hash_ops. We should change it.

Reply:
It does call `hash_ops->free_key` if defined: https://github.com/systemd/systemd/blob/main/src/basic/hashmap.c#L1282

Reply:
Great! Thank you!